### DISCUSSION TITLE
How to apply the registration transform field for a point?

### DISCUSSION THREAD

**Discussion by NeuZhangQiang** [dipy/dipy#2285] [Registration] [answered]
DIPY is really a useful tool . Currently, I use dipy for registration, and the result is good. The code is:

```
sdr = SymmetricDiffeomorphicRegistration(metric, level_iters, ss_sigma_factor=2.5)
static_affine = np.eye(4)
moving_affine = np.eye(4)
pre_align = np.eye(4)
mapping = sdr.optimize(v1, v2, static_affine, moving_affine, pre_align)
```

I know that I can apply the transform field for a image, like:
```
warped_moving = mapping.transform(moving)
```
But, how can I apply the transform field for a point (x, y, z)?

Any suggestion is appreciated!

**Comment by arokem** [comment]
Thanks for your question! The mapping object has 4D volumes called `mapping.forward` and `mapping.backward` that indicate how you would move from a particular position for the forward and backward transform.

So, I think that something like: 

```
x_new = x + mapping.forward[x,y,z][0]
y_new = y + mapping.forward[x,y,z][1]
z_new = z + mapping.forward[x,y,z][2]
```

**Comment by zhang-qiang-github** [comment]
@arokem Thank you very much for your kindly reply. I meet some problem when I transform the point. 

All code and data is updated to [github](https://github.com/zhang-qiang-github/registrationDemo). 

Firstly, I visualize the data and point (showImg.py), and the code is:
```
import numpy as np
from vedo import *  # pip install vedo
data = np.load('img.npz')
img1 = data['img1']
point = [39, 105, 109]
img2 = data['img2']

vol1 = Volume(img1).isosurface(1).c('b').alpha(0.3)
vol2 = Volume(img2).isosurface(1).c('r').alpha(0.3)
p = Point(pos=[point[2], point[1], point[0]])
show(vol1, p, at=0, N=2)
show(vol2, at=1, interactive=True)
```
And the result is:

![image](https://user-images.githubusercontent.com/62534955/98791752-7424ae80-2440-11eb-9b10-72b28e62e267.png)

In the above figure, a red point (```point```) is shown in ```img1```. What I want is map the ```point``` from ```img1``` to ```img2```. 

Then, I perform the SyN registration, and the code is:

```
import numpy as np
from vedo import *  # pip install vedo
data = np.load('img.npz')
img1 = data['img1']
point = [39, 105, 109]
img2 = data['img2']

from dipy.align.imwarp import SymmetricDiffeomorphicRegistration
from dipy.align.imwarp import DiffeomorphicMap
from dipy.align.metrics import CCMetric, SSDMetric
static, moving = img2, img1

metric = SSDMetric(3)
level_iters = [100, 20, 10]
sdr = SymmetricDiffeomorphicRegistration(metric, level_iters, ss_sigma_factor=2.5)
static_affine = np.eye(4)
moving_affine = np.eye(4)
pre_align = np.eye(4)


mapping = sdr.optimize(static, moving, static_affine, moving_affine, pre_align)
warped_moving = mapping.transform(moving)

newX = point[0] + mapping.forward[point[0], point[1], point[2]][0]
newY = point[1] + mapping.forward[point[0], point[1], point[2]][1]
newZ = point[2] + mapping.forward[point[0], point[1], point[2]][2]
p = Point(pos=[newZ, newY, newX])

vol1 = Volume(warped_moving).isosurface(1).c('b').alpha(0.3)
vol2 = Volume(img2).isosurface(1).c('r').alpha(0.3)
p = Point(pos=[point[2], point[1], point[0]])
show(vol1, p, at=0, N=2)
show(vol2, at=1, interactive=True)
```

And the result is:

![image](https://user-images.githubusercontent.com/62534955/98792982-36c12080-2442-11eb-82e5-c2d7e19dbf95.png)

In the above figure, the warp point is on the outside of the warped_moving. But, it should be included by the ```warped_moving```. 

Is there any thing wrong with my code? Your suggestion is really important for me!

**Comment by arokem** [comment]
Not sure, but I would try replacing `mapping.forward` with `mapping.backward`, in case I got the directions wrong.

**Comment by zhang-qiang-github** [comment]
@arokem But，even the mapping.forward is replaced with mapping.backward, the point is still on the outside of warp_img. How can I correctly map the point?

**Comment by arokem** [comment]
I don't know. Sorry.

**Comment by skoudoro** [comment, marked as answer]
Hi @zhang-qiang-github,

We have a function to warp point not yet in DIPY.
For inspiration, you can look at this PR : https://github.com/dipy/dipy/pull/936/files#diff-7e196fd3b79b99d2379639114cbd3fe8ce08c143a597d34714ef88aa94b65de9R293

As you can see, you will have to call some private function but it is doable. Otherwise, you can wait this PR to be merge. We still need to clean and rebase it. 

I hope it helps
